libmisc/idmapping.c

```diff
@@ -132,49 +132,53 @@ void write_mapping(int proc_dir_fd, int ranges, struct map_range *mappings,
 	size_t bufsize;
 	char *buf, *pos;
 	int fd;
-#if HAVE_SYS_CAPABILITY_H
-	struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
-	struct __user_cap_data_struct data[2] = { { 0 } };
-#endif
-
-	bufsize = ranges * ((ULONG_DIGITS  + 1) * 3);
-	pos = buf = xmalloc(bufsize);
 
 #if HAVE_SYS_CAPABILITY_H
+	int cap;
+	struct __user_cap_header_struct hdr = {_LINUX_CAPABILITY_VERSION_3, 0};
+	struct __user_cap_data_struct data[2] = {{0}};
+
+	if (strcmp(map_file, "uid_map") == 0) {
+		cap = CAP_SETUID;
+	} else if (strcmp(map_file, "gid_map") == 0) {
+		cap = CAP_SETGID;
+	} else {
+		fprintf(stderr, _("%s: Invalid map file %s specified\n"), Prog, map_file);
+		exit(EXIT_FAILURE);
+	}
+
 	if (capget(&hdr, data) < 0) {
 		fprintf(stderr, _("%s: Could not get capabilities\n"), Prog);
 		exit(EXIT_FAILURE);
 	}
-	if (!(data[0].effective & CAP_TO_MASK(CAP_SYS_ADMIN)) &&
-	    uid != geteuid()) {
-		bool uid_map;
-
-		if (strcmp(map_file, "uid_map") == 0) {
-			uid_map = true;
-		} else if (strcmp(map_file, "gid_map") == 0) {
-			uid_map = false;
-		} else {
-			fprintf(stderr, _("%s: Invalid map file %s specified\n"), Prog, map_file);
-			exit(EXIT_FAILURE);
-		}
+
+	/* Align setuid- and fscaps-based new{g,u}idmap behavior. */
+	if (!(data[0].effective & CAP_TO_MASK(CAP_SYS_ADMIN)) && uid != 0 &&
```

Comment:
As long as we're doing all this, can we have a comment at top of fn explaining what exactly 'uid' is, or fix up the variable name (maybe caller_ruid, except that can be wrong when actually called by root on behalf of a user) ?

Reply:
done

Reply:
I added an extensive comment now too.